README.md
# NRSurrogate catalog
[![PyPI version](https://badge.fury.io/py/nrsur-catalog.svg)](https://badge.fury.io/py/nrsur-catalog)

This repository contains the python package source code for the NRSurrogate Catalog API.
Install the package with
```
pip install nrsur_catalog
```

This allows users to download catalog data in the CLI, or load results and make plots from within python.


CHANGELOG.md
# Changelog


## V1.0.0
- Release of NRSurCat-1


## V0.0.1
- Pre-release of NRSurCat-1 (API still in development)

src/nrsur_catalog/nrsur_result.py
import contextlib
import os
from typing import List, Optional, Union

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from bilby.gw.result import CompactBinaryCoalescenceResult
from bilby.gw.waveform_generator import WaveformGenerator

from .api import download_event
from .cache import DEFAULT_CACHE_DIR, CatalogCache
from .logger import logger
from .lvk_posterior import load_lvk_result
from .utils import (
    CATALOG_MAIN_COLOR,
    INTERESTING_PARAMETERS,
    LATEX_LABELS,
    get_1d_summary_str,
    get_dir_tree,
    pesummary_to_bilby_result,
    plot_overlaid_corner,
    prior_to_str,
)

pd.set_option("display.max_rows", None, "display.max_columns", None)


class NRsurResult(CompactBinaryCoalescenceResult):
    """Class to interface with a result of a NRSur Catalog analysis."""

    def __init__(self, *args, **kwargs):
        super(NRsurResult, self).__init__(*args, **kwargs)
        self._lvk_result = None

    @classmethod
    def load(
            cls,
            event_name: str,
            cache_dir: Optional[str] = DEFAULT_CACHE_DIR,
            event_path: Optional[str] = None,
    ) -> "NRsurResult":
        """Load a result (either by downloading or from a local dir).

        Parameters
        ----------
        event_name: str
            The name of the event to load
            (list all available events with `nrsur_catalog.catalog.Catalog.
        cache_dir: Optional[str]
            The credible level to use in the plot (default 0.9)
        event_path: Optional[str] = None
            The path to the NRSur result file to load.

        Returns
        -------
        NRsurResult: loaded result object.

        """

        CACHE = CatalogCache(cache_dir)
        if not CACHE.find(event_name) and event_path is None:
            logger.debug(f"{event_name} not in {CACHE.dir}. Downloading...")
            download_event(event_name, cache_dir)

        if event_path is None:
            event_path = CACHE.find(event_name)

        if event_path is None:
            raise ValueError(f"Event {event_name} not found.")

        r = pesummary_to_bilby_result(event_path)
        r.path_to_result = event_path
        r.outdir = os.path.join(CACHE.dir, event_name)
        r.label = event_name
        os.makedirs(r.outdir, exist_ok=True)
        r.__class__ = NRsurResult
        logger.debug(f"Loaded NRresult {event_name} from {event_path}")
        return r

    def summary(self, markdown: bool = False) -> Union[str, pd.DataFrame]:
        """Generate a table of the event parameter summary statistics.

        Parameters
        ----------
        markdown: bool
            True, return a string-markdown table of the summary
            False, return a pd.DataFrame of the summary.
        """

        selected_parameters = []
        summary_str = []
        priors = []

        for _, parameters in INTERESTING_PARAMETERS.items():
            for parameter in parameters:
                if parameter in self.posterior:
                    selected_parameters.append(LATEX_LABELS[parameter])
                    summary_str.append(get_1d_summary_str(self.posterior[parameter]))
                    pri = self.priors.get(parameter, "$-$")
                    if pri != "$-$":
                        pri = prior_to_str(pri)
                    priors.append(pri)

        df = pd.DataFrame(
            {
                "Parameter": selected_parameters,
                "Posterior 90% CI": summary_str,
                "Prior": priors,
            }
        )
        df.index = df["Parameter"]
        df = df.drop(columns=["Parameter"])
        if markdown:
            md_txt = df.to_markdown()
            md_txt = "\n".join([f"# {line}" for line in md_txt.splitlines()])
            return "\n" + md_txt
        else:
            return df

    def plot_skymap(self, *args, **kwargs):
        """Generate a skymap of the event"""
        with open(os.devnull, "w") as f, contextlib.redirect_stdout(f):
            # call the super class sky map method
            super(NRsurResult, self).plot_skymap()

    def __get_waveform_generator(self) -> WaveformGenerator:
        return self.waveform_generator_class(
            duration=self.duration,
            sampling_frequency=self.sampling_frequency,
            start_time=self.start_time,
            frequency_domain_source_model=self.frequency_domain_source_model,
            parameter_conversion=self.parameter_conversion,
            waveform_arguments=self.waveform_arguments,
        )

    def plot_signal(
            self,
            n_samples: Optional[int] = 1000,
            level: Optional[float] = None,
            overplot_max_lnl: Optional[bool] = True,
            polarisation: Optional[str] = "plus",
            outdir: Optional[str] = "",
            overplot_kwargs: Optional[dict] = {},
            kwargs: Optional[dict] = {},
    ) -> plt.Figure:
        """Generate a signal 'trace' plot of the event waveform.

        Parameters
        ----------
        n_samples: int
            The number of samples to use in the plot
        level: float
            The credible level to use in the plot (default None)
        overplot_max_lnl: bool
            If True, overplot the waveform with the maximum log-likelihood
        polarisation: str
            The polarisation to plot
        outdir:str
            The outdir to save the plot. If '' not saved.
        overplot_kwargs: dict
            Kwargs to pass to the overplotted MaxL waveform
        kwargs: dict
            Kwargs to pass to the plot

        """

        if n_samples > len(self.posterior):
            n_samples = len(self.posterior)
            logger.warning(
                "n_samples > len(posterior), using n_samples = len(posterior)"
            )

        samples = self.posterior.sample(n_samples, replace=False)
        samples = samples.reset_index(drop=True)
        base_params = dict(samples.iloc[0])

        # Generate the plotting data
        waveform_generator = self.__get_waveform_generator()

        try:
            # suppress stdout
            with open(os.devnull, "w") as f, contextlib.redirect_stdout(f):
                base_wf = waveform_generator.time_domain_strain(base_params)[
                    polarisation
                ]
        except RuntimeError:
            logger.warning(
                "Unable to create a waveform: do you have NrSur7dq4 installed? Defaulting to IMRPhenomPv2"
            )
            self.waveform_arguments["waveform_approximant"] = "IMRPhenomPv2"
            self.waveform_arguments["minimum_frequency"] = 20
            waveform_generator = self.__get_waveform_generator()
            base_wf = waveform_generator.time_domain_strain(base_params)[polarisation]

        time_idx = np.arange(len(base_wf))
        waveforms = np.zeros((n_samples, len(base_wf)))

        # Max LnL waveform
        max_idx = np.argmax(self.posterior["log_likelihood"])
        max_lnl_params = dict(self.posterior.iloc[max_idx])
        max_lnl_wf = waveform_generator.time_domain_strain(max_lnl_params)[polarisation]

        for i, params in samples.iterrows():
            params = dict(params)
            waveforms[i] = waveform_generator.time_domain_strain(params)[polarisation]
        waveforms = np.roll(waveforms, 4 * len(base_wf) // 5, axis=1)
        max_lnl_wf = np.roll(max_lnl_wf, 4 * len(base_wf) // 5)

        fig, ax = plt.subplots(1, 1, figsize=(6, 5))

        overplot_kwargs["color"] = overplot_kwargs.get("color", CATALOG_MAIN_COLOR)
        overplot_kwargs["alpha"] = overplot_kwargs.get("linewidth", 0.7)
        kwargs["color"] = kwargs.get("color", CATALOG_MAIN_COLOR)
        if level is not None:
            # Calculate the credible region
            delta = (1 + level) / 2
            upper_percentile = delta * 100
            lower_percentile = (1 - delta) * 100
            ci = int(upper_percentile - lower_percentile)
            lower, upper = np.quantile(
                waveforms, [lower_percentile / 100, upper_percentile / 100], axis=0
            )

            kwargs['alpha'] = kwargs.get('alpha', 0.3)
            kwargs['linewidth'] = kwargs.get('linewidth', 0)

            # Plot the data
            ax.fill_between(
                time_idx,
                lower,
                upper,
                label=f"{ci}% credible region",
                **kwargs,
            )
        else:
            kwargs['alpha'] = kwargs.get('alpha', 0.01)
            kwargs['linewidth'] = kwargs.get('linewidth', 0.5)
            for w in waveforms:
                ax.plot(time_idx, w, **kwargs)

        if overplot_max_lnl:
            ax.plot(time_idx, max_lnl_wf, **overplot_kwargs)

        ax.axis("off")
        outdir = self.outdir if outdir == "" else outdir
        filename = os.path.join(outdir, self.label + "_waveform.png")
        fig.tight_layout()
        fig.savefig(filename, dpi=300, bbox_inches="tight", transparent=True)
        return fig

    def plot_corner(
            self,
            parameters: Optional[List[str]] = None,
            priors: Optional[bool] = None,
            titles: Optional[List[str]] = True,
            save: Optional[bool] = False,
            filename: Optional[str] = None,
            dpi: Optional[int] = 300,
            **kwargs,
    ) -> Union[plt.Figure, None]:
        labels = []
        if parameters is not None:
            labels = [LATEX_LABELS[p] for p in parameters]
        if not "labels" in kwargs:
            kwargs["labels"] = labels
        if not "color" in kwargs:
            kwargs["color"] = CATALOG_MAIN_COLOR

        return super(NRsurResult, self).plot_corner(
            parameters, priors, titles, save, filename, dpi, **kwargs
        )

    def plot_lvk_comparison_corner(self, parameters: List[str], **kwargs):
        """Plot a corner plot comparing the posterior to the LVK catalog."""
        return plot_overlaid_corner(
            r1=self,
            r2=self.lvk_result,
            parameters=parameters,
            labels=["NRSur7dq4", "LVK [XPHM]"],
            colors=[CATALOG_MAIN_COLOR, "tab:blue"],
            **kwargs,
        )

    def print_configs(self) -> str:
        configs = self.meta_data["config_file"]
        for key, config in configs.items():
            k = f"{key}:"
            print(f"{k:<35} {config}")

    def download_analysis_datafiles(self, outdir: str):
        """Downloads all the analysis datafiles -- CURRENTLY NOT IMPLEMENTED.
        Please raise a git-issue if this is something you would like us to add.
        """
        if outdir is None:
            outdir = f"outdir_{self.label}"
        raise NotImplementedError(
            "Please raise a git-issue if this is something that you would "
            "really like us to include."
        )
        logger.info(f"Saving files to {outdir}")
        logger.info("To be implemented...")
        logger.info("Downloading analysis strain")
        logger.info("Downloading PSDs")
        logger.info("Downloading calibration files")
        logger.info("writing analysis config file")
        logger.info(f"Files:\n{get_dir_tree(outdir)}")

    @property
    def lvk_result(self) -> CompactBinaryCoalescenceResult:
        """The event's corresponding LVK result (from the GWTC)."""
        if not hasattr(self, "_lvk_result"):
            self._lvk_result = load_lvk_result(
                self.label, cache_dir=os.path.dirname(self.outdir)
            )
            logger.debug(f"Loaded LVKresult{self.label}")
        return self._lvk_result


src/nrsur_catalog/utils/__init__.py
import os
import re
import datetime

import numpy as np
import requests
from bilby.core.prior import Prior
from tqdm.auto import tqdm

from nrsur_catalog.logger import logger
from .constants import (
    CATALOG_MAIN_COLOR,
    INTERESTING_PARAMETERS,
    LATEX_LABELS,
    LOG_PARAMS,
)
from .overlaid_corner import plot_overlaid_corner
from .pesummary_result_to_bilby_result import pesummary_to_bilby_result


def get_size_of_file(filename: str) -> str:
    """Get the size of a file in human-readable format"""
    size = os.path.getsize(filename)
    for x in ["bytes", "KB", "MB", "GB", "TB"]:
        if size < 1024.0:
            return "%3.1f %s" % (size, x)
        size /= 1024.0


def format_qts_to_latex(q_lo: float, q_mid: float, q_hi: float) -> str:
    """Format quantiles to a LaTeX string"""
    q_m, q_p = q_mid - q_lo, q_hi - q_mid
    fmt = "{{0:{0}}}".format(".2f").format
    summary = r"${{{0}}}_{{-{1}}}^{{+{2}}}$"
    return summary.format(fmt(q_mid), fmt(q_m), fmt(q_p))


def get_1d_summary_str(x: np.ndarray, quantiles=[0.16, 0.5, 0.84]) -> str:
    q_lo, q_mid, q_hi = np.quantile(x, quantiles)
    return format_qts_to_latex(q_lo, q_mid, q_hi)


def get_event_name(s: str):
    """Get the event name from a string using a regex
    e.g.
    'GW170817' from 'GW170817_4.5_0.1.hdf5'
    'GW150914' from "https://sandbox.zenodo.org/record/1164558/files/GW150914_result.json"
    'GW200224_222234'  from  "GW200224_222234_NRSur7dq4_merged_result.json"
    """
    try:
        return re.findall(r"(GW\d{6}\_\d{6}|GW\d{6})", s)[0]
    except IndexError:
        logger.debug(f"Could not parse event name from {s}")
        return None

def get_gps_time(s: str)->int:
    """
    eg G150914_095045 --> GPS(2015-09-14T09:50:45)
    """
    t1 = datetime.datetime.strptime(s, "GW%y%m%d_%H%M%S")
    t0 = datetime.datetime(1980, 1, 6)
    return int((t1-t0).total_seconds())

def get_dir_tree(path):
    """Get a directory tree (in a string)"""
    tree = ""
    for root, dirs, files in os.walk(path):
        level = root.replace(path, "").count(os.sep)
        indent = " " * 4 * (level)
        tree += f"{indent}{os.path.basename(root)}/\n"
        subindent = " " * 4 * (level + 1)
        for f in files:
            tree += f"{subindent}{f}\n"
    return tree


def download(url: str, fname: str) -> None:
    """Download a file from a URL and save it to a file"""
    logger.info(f"<<Downloading {url} to {fname}>>")
    response = requests.get(url, stream=True)
    total_size = int(response.headers.get('content-length', 0))

    basefn = os.path.basename(fname)
    with open(fname, 'wb') as file, tqdm(
            desc=basefn,
            total=total_size,
            unit='B',
            unit_scale=True,
            unit_divisor=1024,
            ncols=100,
    ) as bar:
        for data in response.iter_content(chunk_size=1024):
            size = file.write(data)
            bar.update(size)


def prior_to_str(prior: Prior):
    """Convert a prior to a string"""
    # get class name
    name = prior.__class__.__name__
    min_val, max_val = None, None
    if hasattr(prior, "minimum"):
        min_val = round(prior.minimum, 2)
    if hasattr(prior, "maximum"):
        max_val = round(prior.maximum, 2)
    repr = r"$\text{" + name + "}"
    if min_val is not None and max_val is not None:
        repr += f" [{min_val}, {max_val}]"
    return repr + "$"


def safe_plot(func):
    """Decorator to catch errors when plotting"""

    def wrapper(*args, **kwargs):
        try:
            func(*args, **kwargs)
        except Exception as e:
            logger.error(f"Failed to plot {func.__name__}: {e}")

    return wrapper


src/nrsur_catalog/utils/constants.py
INTERESTING_PARAMETERS = {
    "Mass Parameters": ["mass_1_source", "mass_2_source", "chirp_mass", "mass_ratio"],
    "Spin Parameters": ["a_1", "a_2", "tilt_1", "tilt_2", "chi_eff", "chi_p"],
    "Localisation Parameters": [
        "ra",
        "dec",
        "geocent_time",
        "luminosity_distance",
    ],
    "Other Parameters": [
        "phase",
        "azimuth",
        "zenith",
        "psi",
        "phi_jl",
        "phi_12",
        "theta_jn",
    ],
    "Remnant Parameters": ["final_mass", "final_spin", "final_kick"],
}


CATALOG_MAIN_COLOR = "tab:orange"

LATEX_LABELS = dict(
    mass_1=r"$m_1^{\rm det}\ [M_{\odot}]$",
    mass_2=r"$m_2^{\rm det}\ [M_{\odot}]$",
    mass_1_source=r"$m_1^{\rm src}\ [M_{\odot}]$",
    mass_2_source=r"$m_2^{\rm src}\ [M_{\odot}]$",
    chirp_mass=r"$\mathcal{M}\ [M_{\odot}]$",
    mass_ratio=r"$q$",
    a_1=r"$a_1$",
    a_2=r"$a_2$",
    tilt_1=r"$\theta_1$",
    tilt_2=r"$\theta_2$",
    chi_eff=r"$\chi_{\mathrm{eff}}$",
    chi_p=r"$\chi_{\mathrm{p}}$",
    ra=r"$\alpha$",
    dec=r"$\delta$",
    geocent_time=r"$t_c\ [s]$",
    luminosity_distance=r"$d_L\ [Mpc]$",
    phase=r"$\phi$",
    azimuth=r"$\phi$",
    zenith=r"$\theta$",
    psi=r"$\psi$",
    phi_jl=r"$\phi_{\mathrm{JL}}$",
    phi_12=r"$\phi_{12}$",
    theta_jn=r"$\theta_{\mathrm{JN}}$",
    final_mass=r"$m_f\ [M_{\odot}]$",
    final_spin=r"$\chi_f$",
    final_kick=r"$v_{\rm kick}\ [{\rm km/s}]$",
)


LOG_PARAMS = ["mass_1", "mass_2"]


src/nrsur_catalog/catalog.py
"""Module to load all the catalog events and make plots with them"""
import os.path
from glob import glob
from itertools import chain
from typing import Dict, List, Optional, Union

import h5py
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from corner import hist2d
from matplotlib.ticker import AutoMinorLocator
from tqdm.auto import tqdm

from .api import download_missing_events, get_analysed_event_names
from .cache import DEFAULT_CACHE_DIR, NR_FILE_EXTENSION, CatalogCache
from .logger import logger
from .nrsur_result import NRsurResult
from .utils import (
    CATALOG_MAIN_COLOR,
    INTERESTING_PARAMETERS,
    LATEX_LABELS,
    format_qts_to_latex,
    get_event_name,
    get_gps_time
)

POSTERIORS_TO_KEEP = list(
    chain.from_iterable(v for v in INTERESTING_PARAMETERS.values())
)
# POSTERIORS_TO_KEEP += ["log_likelihood", "log_prior"]
MAX_SAMPLES = 10000
CATALOG_FN = "downsampled_posteriors.h5"

DEFAULT_CLEAN_CATALOG = False


class Catalog:
    def __init__(self, posteriors: pd.DataFrame):
        """Class to store the catalog events

        Parameters
        -----------
        posteriors: pd.DataFrame
            A dataframe with the catalog posteriors
            The dataframe must have an `event` column with the event name
        """
        self._df = posteriors
        # sort by event name (GW150914, GW151012, ...)
        self._df = self._df.sort_values("event")

    @classmethod
    def load(
        cls,
        cache_dir: Optional[str] = DEFAULT_CACHE_DIR,
        max_samples: Optional[int] = MAX_SAMPLES,
        clean: Optional[bool] = DEFAULT_CLEAN_CATALOG,
    ) -> "Catalog":
        """Load all the catalog posterior samples and stores a cache of the down-sampled posteriors.

        Parameters
        ----------
        cache_dir: Optional[str]
            The cache-dir where the NRSurrogate results will be loaded from.
            If empty, downloads all the NRSurrogate results.
        max_samples: Optional[int]
            The max number of samples to include from each event's posterior.
        clean: Optional[bool]
            Recomputes the cache of the downsampled-posteriors.
        """
        CACHE = CatalogCache(cache_dir)
        fname = f"{cache_dir}/{CATALOG_FN}"

        if not clean and os.path.isfile(fname):
            logger.info(f"Loading catalog from {fname}")
            catalog = cls(Catalog.from_hdf5(fname))
        else:
            if CACHE.is_incomplete:
                logger.warning(
                    f"Cache {cache_dir} is missing some events [{CACHE.missing_events}], downloading events..."
                )
                download_missing_events(CACHE.dir)
            df = Catalog.__load_events(cache_dir, max_samples=max_samples)
            catalog = cls(df)
            catalog.save(f"{cache_dir}/{CATALOG_FN}")
        return catalog

    @staticmethod
    def __load_events(events_dir: str, max_samples: int = MAX_SAMPLES) -> pd.DataFrame:
        """Helped method to load the posteriors from the Cache dir."""
        event_paths = glob(f"{events_dir}/*{NR_FILE_EXTENSION}")
        dfs = []
        assert len(event_paths) > 0, f"No events found in {events_dir}"
        event_paths = sorted(event_paths, key=get_event_name)
        for path in tqdm(event_paths, desc="Loading events"):
            event_name = get_event_name(path)

            r = NRsurResult.load(event_name, event_path=path)
            missing_params = set(POSTERIORS_TO_KEEP) - set(r.posterior.columns)
            if len(missing_params) > 0:
                logger.warning(
                    f"Event {event_name} does not have all the required parameters. Missing:"
                    f"{missing_params}. Adding mock data..."
                )
                for p in missing_params:
                    r.posterior[p] = np.nan
            posterior = r.posterior[POSTERIORS_TO_KEEP]
            if len(r.posterior) > max_samples:
                logger.debug(
                    f"Event {event_name} has {len(r.posterior)}>{max_samples} samples, "
                    f"downsampling to {max_samples} samples"
                )
                posterior = posterior.sample(
                    max_samples, weights=r.posterior.log_likelihood
                )
            posterior.loc[:, "event"] = event_name
            dfs.append(posterior)
        return pd.concat(dfs)

    def to_dict_of_posteriors(
        self, parameters: Optional[List[str]] = None
    ) -> Dict[str, pd.DataFrame]:
        """Return a dictionary of event:posteriors.

        Parameters
        ----------
        parameters: Optional[List[str]]
            List of the parameters to include in the posterior.


        Returns
        -------
        Dict[str:pd.DataFrame]: Returns a dictionary of {events: event_posteriors}.

        """
        if parameters is None:
            parameters = POSTERIORS_TO_KEEP
        return {
            e: self._df[self._df["event"] == e][parameters] for e in self.event_names
        }

    def save(self, path: str) -> None:
        """Cache the loaded catalog to a hdf5 file.

        Parameters
        ---------
        path: str
            Path to save the cache-catalog

        """
        assert path.endswith(".h5"), "Path must end with .h5"
        data = self.to_dict_of_posteriors()
        with h5py.File(path, "w") as f:
            for event_name, posterior in data.items():
                # write each posterior to a different key in the hdf5 file
                f.create_dataset(event_name, data=posterior.to_records(index=False))

    @staticmethod
    def from_hdf5(path: str) -> pd.DataFrame:
        """Load the catalog from a hdf5 file"""
        assert path.endswith(".h5"), "Path must end with .h5"
        with h5py.File(path, "r") as f:
            posteriors = []
            for key in f:
                df = pd.DataFrame(f[key][:])
                df["event"] = key
                posteriors.append(df)
            return pd.concat(posteriors)

    def violin_plot(self, parameter: str, sort_by='event_name') -> plt.Figure:
        """Generate a violin plot of a parameter.

        Parameters
        ----------
        parameter: str
            The parameter to plot.
        sort_by: str
            Either 'median' or 'event_name'.
            If 'median', the events will be sorted by the median value of the parameter.
            If 'event_name', the events will be sorted chronologically by their name.
        """
        # posterior subsets for each event


        if sort_by == 'event_name':
            ylabels = self.event_names[::-1]
            data = self.to_dict_of_posteriors([parameter])
            data = [data[name][parameter].values.ravel() for name in ylabels]
        elif sort_by == 'median':
            data = [
                post.values.ravel()
                for post in self.to_dict_of_posteriors([parameter]).values()
            ]
            medians = [np.mean(d) for d in data]
            ylabels = [e for _, e in sorted(zip(medians, self.event_names))]
            data = [d for _, d in sorted(zip(medians, data), key=lambda x: x[0])]
        else:
            raise ValueError(f"sort_by must be 'event_name' or 'median', not {sort_by}")

        fig, ax = plt.subplots(figsize=(7, self.n * 0.8))
        ax.tick_params(axis="x", bottom=True, top=True, labelbottom=True, labeltop=True)
        for pos in ["right", "top", "bottom", "left"]:
            ax.spines[pos].set_visible(False)
        ax.set_xticks([])
        ax.set_title(LATEX_LABELS[parameter], fontsize=22)
        ax.set_xlabel(LATEX_LABELS[parameter], fontsize=22)
        ax.grid(True, axis="x", alpha=0.25)
        ax.xaxis.set_tick_params(width=0)
        ax.yaxis.set_tick_params(width=0)
        ax.set_yticks(range(1, self.n + 1))
        ax.set_yticklabels(ylabels)
        violin_parts = ax.violinplot(data, vert=False)
        for pc in violin_parts["bodies"]:
            pc.set_color(CATALOG_MAIN_COLOR)
        for partname in ("cbars", "cmins", "cmaxes"):
            vp = violin_parts[partname]
            vp.set_edgecolor(CATALOG_MAIN_COLOR)
        min_x = min([min(d) for d in data])
        max_x = max([max(d) for d in data])
        axes_ticks = np.linspace(min_x, max_x, 5)
        ax.set_ylim(0.5, self.n + 0.5)
        ax.set_xticks(axes_ticks)

        # annotate a group number every 7 y ticks
        ntics = 7
        for i in range(6, self.n, ntics):
            for xval in axes_ticks:
                ax.text(
                    xval,
                    i + 1.5,
                    f"{xval:.1f}",
                    fontsize=12,
                    ha="center",
                    va="center",
                    color="black",
                    alpha=0.25,
                )

        plt.tight_layout()
        fig.savefig(f"{parameter}_violin.png", dpi=300, bbox_inches="tight")
        return fig

    @property
    def parameters(self) -> List[str]:
        """Return a list of parameters in the catalog"""
        p = list(self._df.columns)
        p.remove("event")
        return p

    def get_posterior_quantiles(self, quantiles=[0.16, 0.5, 0.84]) -> pd.DataFrame:
        """Return posterior quantiles for each event's posteriors.

        Parameters
        ----------
        quantiles: List of floats marking the quantiles.

        """
        quant_df = self._df.groupby("event").quantile(quantiles)
        quant_df.index = quant_df.index.rename(["event", "quantile"])
        return quant_df

    @property
    def event_names(self) -> List[str]:
        """Return a list of loaded events in the catalog object (sorted chronologically)"""
        names = list(self._df.event.unique())
        return sorted(names, key=lambda x: get_gps_time(x))

    @property
    def n(self) -> int:
        """Return the number of events in the catalog"""
        return len(self.event_names)

    def get_latex_summary(self) -> pd.DataFrame:
        """
        Return a dataframe with the median and 68% credible interval for each parameter
        """
        post_quantiles = self.get_posterior_quantiles()
        latex_summary = {}
        for event in self.event_names:
            event_data = {}
            for param in self.parameters:
                event_quants = post_quantiles.loc[event, param].values
                event_data[param] = format_qts_to_latex(*event_quants)
            latex_summary[event] = event_data

        latex_summary = pd.DataFrame(latex_summary).T
        latex_summary.index.name = "event"
        return latex_summary

    def plot_2d_posterior(
        self,
        parm_x: str,
        parm_y: str,
        scatter_medians: Optional[bool] = True,
        event_posteriors: Optional[bool] = False,
        event_quantiles: Optional[bool] = True,
        contour_all_posterior_samples: Optional[bool] = True,
        colors: Optional[Union[str, List[str]]] = CATALOG_MAIN_COLOR,
    ) -> plt.Figure:
        """
        Plot a 2D posterior for all events in the catalog

        Parameters
        ----------
        parm_x:str
            The parameter plotted along the x-axis (e.g. mass_1, chi_eff...)
        parm_y:str
            The parameter plotted along the y-axis (e.g. mass_1, chi_eff...)
        scatter_medians:Optional[bool]=True
            Scatter median values of each event.
        event_posteriors:Optional[bool]=False
            Outline the event posterior.
        event_quantiles:Optional[bool]=True
            Plot the 68% CI quantiles of each event's posterior.
        contour_all_posterior_samples:Optional[bool]=True
            Plot contours around all the posterior samples
            (regardless of which event each sample belongs to)
        colors:Optional[Union[str, List[str]]]
            The color to be used for each event.
            If a string, the same color will be used for each event.

        Returns
        -------
        plt.Figure: A matplotlib figure with the plot.
        """
        fig, ax = plt.subplots(figsize=(5, 5))
        if isinstance(colors, str):
            colors = [colors] * self.n

        if event_posteriors or contour_all_posterior_samples:
            if event_posteriors:
                linewidths = 0.2
            else:
                linewidths = 0

            for i, event in enumerate(self.event_names):
                data = self.get_event_posterior(event)[[parm_x, parm_y]]
                hist2d(
                    ax=ax,
                    x=data[parm_x].values,
                    y=data[parm_y].values,
                    color=colors[i],
                    bins=50,
                    zorder=-2,
                    alpha=0.1,
                    plot_datapoints=False,
                    plot_density=False,
                    plot_contours=True,
                    no_fill_contours=False,
                    fill_contours=True,
                    levels=[0, 1.0 - np.exp(-0.5 * 2.1 ** 2)],
                    smooth=1.1,
                    contourf_kwargs=dict(linewidths=linewidths),
                    contour_kwargs=dict(linewidths=linewidths),
                )

        if scatter_medians:
            posterior_quantiles = self.get_posterior_quantiles()
            posterior_quantiles = posterior_quantiles[[parm_x, parm_y]]
            for i, event in enumerate(self.event_names):
                x, y = posterior_quantiles.loc[event].values[1]
                xlow, ylow = posterior_quantiles.loc[event].values[0]
                xhigh, yhigh = posterior_quantiles.loc[event].values[2]
                ax.scatter(x, y, color=colors[i], s=0.5)
                if event_quantiles:
                    ax.errorbar(
                        x,
                        y,
                        xerr=[[x - xlow], [xhigh - x]],
                        yerr=[[y - ylow], [yhigh - y]],
                        fmt="none",
                        color=colors[i],
                        alpha=0.2,
                        capsize=0,
                    )

        # add minor ticks
        ax.set_xlim(min(self._df[parm_x]), max(self._df[parm_x]))
        ax.set_ylim(min(self._df[parm_y]), max(self._df[parm_y]))
        ax.tick_params(
            axis="both",
            which="both",
            direction="in",
            top=True,
            right=True,
            labelsize=12,
            pad=5,
        )
        ax.tick_params(axis="x", pad=10)
        # add minor ticks
        ax.xaxis.set_minor_locator(AutoMinorLocator())
        ax.yaxis.set_minor_locator(AutoMinorLocator())
        ax.tick_params(axis="both", which="major", length=8)
        ax.tick_params(axis="both", which="minor", length=4)
        ax.set_xlabel(LATEX_LABELS[parm_x])
        ax.set_ylabel(LATEX_LABELS[parm_y])
        plt.tight_layout()
        return fig

    def get_all_posteriors(self) -> pd.DataFrame:
        """Return the dataframe of all the event posteriors concatenated together.
        The column 'event' denotes which 'event' the posterior sample belongs to.

        Returns
        -------
        pd.DataFrame: A dataframe of the all event's posterior samples.
        """
        return self._df.copy()

    def get_event_posterior(self, event_name: str) -> pd.DataFrame:
        """
        Return a dataframe with the posterior for a given event.

        Parameter
        ---------
        event_name:str
            The name of the event that will have its posterior returned.

        Returns
        -------
        pd.DataFrame: A dataframe of the event's posterior samples
        """
        return self._df[self._df["event"] == event_name]

    @staticmethod
    def get_analysed_event_names() -> List[str]:
        """Return the list of the analysed events available in the NRSurrogate Catalog.

        Returns
        -------
        List[str]: List of the analysed event names.
        """
        return get_analysed_event_names()

    @property
    def all_analysed_events_loaded(self) -> bool:
        """True if all analysed events have been loaded
        (this will be False when the Catalog object is created
        before all event results have been downloaded).
        """
        analysed_events = set(self.get_analysed_event_names())
        loaded_events = set(self.event_names)
        return len(analysed_events - loaded_events) == 0
